File: frontend/src/api/webhooks.ts
/**
 * Webhooks API client (v1.3 S8-2). Admin-only CRUD + test over the
 * admin-global webhook endpoints. The server fires deliveries from its
 * events-bus subscriber; this client only manages config.
 */
import { fetchJSON, fetchURL } from "./utils";

export interface WebhookEndpoint {
  id: string;
  url: string;
  enabled: boolean;
  /** Per-endpoint event-type filter; empty = all file events. */
  events: string[];
  lastStatus?: string; // "" | "success" | "failed"
  lastCode?: number;
  lastError?: string;
  lastAt?: string;
}

export interface WebhookListResponse {
  endpoints: WebhookEndpoint[];
  eventTypes: string[];
}

export interface WebhookInput {
  url: string;
  enabled: boolean;
  events: string[];
}

export interface WebhookTestResult {
  ok: boolean;
  code: number;
  error?: string;
}

export async function list(): Promise<WebhookListResponse> {
  return fetchJSON<WebhookListResponse>("/api/webhooks", {});
}

export async function create(input: WebhookInput): Promise<WebhookEndpoint> {
  return fetchJSON<WebhookEndpoint>("/api/webhooks", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function update(
  id: string,
  input: WebhookInput
): Promise<WebhookEndpoint> {
  return fetchJSON<WebhookEndpoint>(`/api/webhooks/${id}`, {
    method: "PUT",
    body: JSON.stringify(input),
  });
}

export async function remove(id: string): Promise<void> {
  await fetchURL(`/api/webhooks/${id}`, { method: "DELETE" });
}

export async function test(id: string): Promise<WebhookTestResult> {
  return fetchJSON<WebhookTestResult>(`/api/webhooks/${id}/test`, {
    method: "POST",
  });
}

File: http/webhooks.go
package fbhttp

import (
	"encoding/json"
	"errors"
	"net/http"
	"strings"

	"github.com/gorilla/mux"

	"github.com/filebrowser/filebrowser/v2/webhooks"
)

// webhookBody is the editable shape of an endpoint (PUT/POST). Status
// fields are server-managed and not accepted from the client.
type webhookBody struct {
	URL     string   `json:"url"`
	Enabled bool     `json:"enabled"`
	Events  []string `json:"events"`
}

func validWebhookURL(u string) bool {
	return strings.HasPrefix(u, "http://") || strings.HasPrefix(u, "https://")
}

// sanitizeEvents drops anything that isn't a known file-event type. An
// empty result means "all events" (the dispatcher treats it that way).
func sanitizeEvents(in []string) []string {
	valid := map[string]bool{}
	for _, t := range webhooks.FileEventTypes() {
		valid[t] = true
	}
	out := []string{}
	for _, t := range in {
		if valid[t] {
			out = append(out, t)
		}
	}
	return out
}

func webhooksListHandler(store *webhooks.Store) handleFunc {
	return withAdmin(func(w http.ResponseWriter, r *http.Request, _ *data) (int, error) {
		eps, err := store.List()
		if err != nil {
			return http.StatusInternalServerError, err
		}
		if eps == nil {
			eps = []webhooks.Endpoint{}
		}
		return renderJSON(w, r, map[string]interface{}{
			"endpoints":  eps,
			"eventTypes": webhooks.FileEventTypes(),
		})
	})
}

func webhooksCreateHandler(store *webhooks.Store) handleFunc {
	return withAdmin(func(w http.ResponseWriter, r *http.Request, _ *data) (int, error) {
		var body webhookBody
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			return http.StatusBadRequest, err
		}
		if !validWebhookURL(body.URL) {
			return http.StatusBadRequest, errors.New("webhook url must be http(s)")
		}
		ep := &webhooks.Endpoint{
			URL:     body.URL,
			Enabled: body.Enabled,
			Events:  sanitizeEvents(body.Events),
		}
		if err := store.Create(ep); err != nil {
			return http.StatusInternalServerError, err
		}
		return renderJSON(w, r, ep)
	})
}

func webhooksUpdateHandler(store *webhooks.Store) handleFunc {
	return withAdmin(func(w http.ResponseWriter, r *http.Request, _ *data) (int, error) {
		var body webhookBody
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			return http.StatusBadRequest, err
		}
		if !validWebhookURL(body.URL) {
			return http.StatusBadRequest, errors.New("webhook url must be http(s)")
		}
		ep := &webhooks.Endpoint{
			ID:      mux.Vars(r)["id"],
			URL:     body.URL,
			Enabled: body.Enabled,
			Events:  sanitizeEvents(body.Events),
		}
		if err := store.Update(ep); err != nil {
			if errors.Is(err, webhooks.ErrNotFound) {
				return http.StatusNotFound, nil
			}
			return http.StatusInternalServerError, err
		}
		return renderJSON(w, r, ep)
	})
}

func webhooksDeleteHandler(store *webhooks.Store) handleFunc {
	return withAdmin(func(_ http.ResponseWriter, r *http.Request, _ *data) (int, error) {
		if err := store.Delete(mux.Vars(r)["id"]); err != nil {
			return http.StatusInternalServerError, err
		}
		return http.StatusOK, nil
	})
}

// webhooksTestHandler fires a synthetic `webhook.test` POST at the saved
// endpoint and reports the result (also recorded as the endpoint's last
// delivery). Synchronous — the admin is waiting on the result.
func webhooksTestHandler(store *webhooks.Store, dispatcher *webhooks.Dispatcher) handleFunc {
	return withAdmin(func(w http.ResponseWriter, r *http.Request, _ *data) (int, error) {
		ep, err := store.Get(mux.Vars(r)["id"])
		if err != nil {
			return http.StatusInternalServerError, err
		}
		if ep == nil {
			return http.StatusNotFound, nil
		}
		code, derr := dispatcher.Test(ep.ID, ep.URL)
		result := map[string]interface{}{
			"ok":   derr == nil && code >= 200 && code < 300,
			"code": code,
		}
		if derr != nil {
			result["error"] = derr.Error()
		}
		return renderJSON(w, r, result)
	})
}
